[Arm64/Unix] Use portable MP optimized new/alloc

sdmaclea opened the pull request.

sdmaclea commented:
@arm64-contrib @jkotas @janvorli This is a big win on MP ARM64.  Total elapsed test time to run 11K tests reduced ~33%.

sdmaclea commented:
Code is largely copied from `jitinterfacegen.cpp`'s version of `void InitJITHelpers1()`

jkotas commented on src/vm/arm64/stubs.cpp:
```diff
@@ -1086,6 +1087,31 @@ void JIT_TailCall()
+
+    _ASSERTE(g_SystemInfo.dwNumberOfProcessors != 0);
+
+    // TODO-ARM64-NYI !defined(FEATURE_IMPLICIT_TLS) Use helpers optimized w/ inline TLS access...
```
This TODO is not necessary. `FEATURE_IMPLICIT_TLS` is as good as what you can get the handcrafter helpers.

sdmaclea replied:
The comment is for the Windows !`FEATURE_IMPLICIT_TLS` case.

sdmaclea replied:
It looks like `FEATURE_IMPLICIT_TLS` is not defined for Windows.  Am I wrong?

janvorli replied:
In the root clrdefinitions.cmake:
```cmake
if (CLR_CMAKE_PLATFORM_UNIX OR CLR_CMAKE_TARGET_ARCH_ARM64)
  add_definitions(-DFEATURE_IMPLICIT_TLS)
  set(FEATURE_IMPLICIT_TLS 1)
endif(CLR_CMAKE_PLATFORM_UNIX OR CLR_CMAKE_TARGET_ARCH_ARM64)
```
So for ARM64, it is defined for both Windows and Unix.

sdmaclea replied:
Done.  Closed #12332

sdmaclea commented:
@jkotas @janvorli This is ready to merge

jkotas commented:
@jashook Could you please trigger whatever arm64 testing is appropriate?

jashook commented:
Currently testing is offline and I am looking into the issue. If I cannot find the solution in a reasonable amount of time it is probably worth not blocking.

sdmaclea commented:
FYI, I did run this through Unix/Arm64 testing locally.

sdmaclea commented:
@dotnet-bot test Ubuntu arm64 Cross Debug Build

sdmaclea commented:
ping

jashook commented:
@dotnet-bot test Windows_NT Arm64 Checked

jashook commented:
Thank you for waiting @sdmaclea

jashook commented:
@sdmaclea looks like there is a contract violation

sdmaclea commented:
@jashook What does a contract violation mean?  I have seen these CONTRACT things floating around, but I haven't figured out the syntax or the tool used to check them.

jkotas commented:
The CONTRACT things are described in https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#2.10

They get checked as asserts at runtime - on Windows only. The contract violation is basically a special form of assertion failure.

I have fixed a contract violation in https://github.com/dotnet/coreclr/pull/12542. Was this change included in the run?

jashook commented:
It is hard to say, I can restart the job to to see if it repros now.

@dotnet-bot test Windows_NT Arm64 Checked

sdmaclea commented:
Build timeout
@dotnet-bot test Windows_NT Arm64 Checked

sdmaclea commented:
@jkotas Contract violation occurs with #12542.  My guess is the CV is in the arm64 tip.  Triggered "Windows_NT arm64 Cross Checked Build and Test" on #12436 to test hypothesis.

jashook commented:
@sdmaclea Windows arm64 jobs should be clean currently.

sdmaclea commented:
@jashook @jkotas Confirming #12436 completed w/o contract violation.

sdmaclea commented:
@dotnet-bot test Ubuntu arm64 Cross Debug Build

sdmaclea commented:
@jashook FWIW Looks like arm64_debug_small_page_size_prtest pass, but arm64_debug_small_page_size_flow_prtest do not.
+ #12334 passed https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_debug_small_page_size_prtest/250/
+ #12333 failed
https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm64_debug_small_page_size_flow_prtest/173/

sdmaclea commented:
@jkotas I removed the contract line I added.  Since I do not have a Windows Arm64 platform, this the CV is not something I can easily debug.  I'll open an issue to add the contract.

sdmaclea commented:
@jkotas @dotnet/arm64-contrib please review and/or merge

jkotas commented:
@dotnet/arm64-contrib Are Windows amd64 tests passing on this change?

I see that `Windows_NT arm64 Cross Debug Build` is green, but I do not see any tests run in the log.

> Remove change to InitJITHelpers1() contract

IIRC, contracts that failed on this earlier were not related to InitJITHelpers1. I do not think that there was anything wrong with the InitJITHelpers1 contract that you have added. It is ok to look into it independently since you have filled issue on it.

sdmaclea commented:
@dotnet-bot test Windows_NT arm64 Checked

sdmaclea commented:
@jkotas Since I do not have Windows test platform to debug Contract violation, I disabled the change for Windows and changed #13053 to match.

sdmaclea commented:
@jkotas 
> I assume that you have done a test pass to verify that there are not regressions on Linux.

Yes.  I have been using the patch for all my runs for the last two months.  Either way, I will try to get someone form the QDT windows team to debug the Contract violation.

> The regression will likely show up as a intermittent crash during GC.

Will GC stress testing help flush it out?

arm64 GC is not perfect yet.  there are still interrmittent failures.  So It may not be obvious that it is a CV related issue.

jkotas commented:
> I have been using the patch for all my runs for the last two months

This should be good enough.

> Will GC stress testing help flush it out?

Maybe. It is best to debug the known problem on Windows first, I think.

sdmaclea commented:
@dotnet-bot test Ubuntu arm64 Cross Debug Build

sdmaclea commented:
Please merge

sdmaclea commented on src/vm/arm64/stubs.cpp:
```diff
@@ -1086,6 +1087,31 @@ void JIT_TailCall()
 #if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
 void InitJITHelpers1()
 {
+#ifdef FEATURE_PAL // TODO
```
This should be removed to support Windows
